File: js-node/run-tests.js
"use strict";

// A command-line program that runs the dropbox-content-hasher tests.

const crypto = require('crypto');
const dch = require('./dropbox-content-hasher');

function referenceHasher(data) {
  const overallHasher = crypto.createHash('sha256');
  for (let pos = 0; pos < data.length; pos += dch.BLOCK_SIZE) {
    const chunk = data.slice(pos, pos+dch.BLOCK_SIZE);
    const blockHasher = crypto.createHash('sha256');
    blockHasher.update(chunk);
    overallHasher.update(blockHasher.digest());
  }
  return overallHasher.digest('hex');
}

function check(data, chunkSizes) {
  console.log("checking " + JSON.stringify(chunkSizes));

  const hasher = dch.create();

  let pos = 0;
  for (const chunkSize of chunkSizes) {
    const chunk = data.slice(pos, pos+chunkSize);
    pos += chunkSize;

    hasher.update(chunk);
  }

  const result = hasher.digest('hex');
  const reference = referenceHasher(data.slice(0, sum(chunkSizes)))

  const passed = (result === reference);
  if (!passed) {
    console.log("- FAILED: " + JSON.stringify(reference) + ", " + JSON.stringify(result));
  }
  return passed;
}

function sum(arr) {
  let r = 0;
  for (const n of arr) {
    r += n;
  }
  return r;
}

function main() {
  const args = process.argv.slice(2);
  if (args.length > 0) {
    console.error("No arguments expected; got " + args.length + ".");
    process.exit(1);
  }

  const B = dch.BLOCK_SIZE;

  const tests = [
    [0],
    [100],
    [100, 10],
    [B-1],
    [B],
    [B+1],

    [B-2, 1],
    [B-2, 2],
    [B-2, 3],

    [B-2, B+1],
    [B-2, B+2],
    [B-2, B+3],

    [5, 5, 5],
    [5, 5, 5, B],
    [5, 5, 5, 3*B],
    [5, 5, 5, 3*B, 5, 5, 5, 3*B],
  ];

  let longestLength = 0;
  for (const test of tests) {
    longestLength = Math.max(longestLength, sum(test));
  }

  console.log("generating random data");
  const data = crypto.randomBytes(longestLength);

  for (const test of tests) {
    const passed = check(data, test);
    if (!passed) {
      process.exit(2);
    }
  }
  console.log("all passed");
}

main();

File: rust/src/run_tests.rs
//! A command-line tool that runs the `DropboxContentHasher` tests.

extern crate digest;
extern crate rand;
extern crate dropbox_content_hasher;
extern crate sha2;

use digest::Digest;
use std::io::Write as io_Write;
use rand::Rng;
use dropbox_content_hasher::{DropboxContentHasher, BLOCK_SIZE, hex};

fn main() {
    let mut args = std::env::args();
    args.next().unwrap();  // Remove name of binary.
    if args.len() != 0 {
        writeln!(&mut std::io::stderr(), "No arguments expected; got {}.", args.len()).unwrap();
        std::process::exit(1);
    }

    let b = BLOCK_SIZE;

    let tests: &[&[usize]] = &[
        &[0],
        &[100],
        &[100, 10],
        &[b-1],
        &[b],
        &[b+1],

        &[b-2, 1],
        &[b-2, 2],
        &[b-2, 3],

        &[b-2, b+1],
        &[b-2, b+2],
        &[b-2, b+3],

        &[5, 5, 5],
        &[5, 5, 5, b],
        &[5, 5, 5, 3*b],
        &[5, 5, 5, 3*b, 5, 5, 5, 3*b],
    ];

    let longest_length = tests.iter().fold(0, |m, x| std::cmp::max(m, x.iter().sum()));

    println!("generating random data");
    let mut data: Box<[u8]> = vec![0; longest_length].into_boxed_slice();
    rand::ChaChaRng::new_unseeded().fill_bytes(data.as_mut());

    for &test in tests {
        let passed = check(data.as_ref(), test);
        if !passed {
            std::process::exit(2);
        }
    }
    println!("all passed");
}

fn reference_hasher(data: &[u8]) -> String {
    let mut overall_hasher = sha2::Sha256::new();
    for chunk in data.chunks(BLOCK_SIZE) {
        let mut block_hasher = sha2::Sha256::new();
        block_hasher.input(chunk);
        overall_hasher.input(block_hasher.result().as_slice());
    }
    hex(overall_hasher.result().as_slice())
}

fn check(data: &[u8], chunk_sizes: &[usize]) -> bool {
    println!("checking {:?}", chunk_sizes);

    let mut hasher = DropboxContentHasher::new();

    let mut input = data;
    let mut total_length = 0;
    for chunk_size in chunk_sizes.iter().cloned() {
        let (chunk, rest) = input.split_at(chunk_size);
        input = rest;
        hasher.input(chunk);
        total_length += chunk_size;
    }

    let result = hex(hasher.result().as_slice());
    let reference = reference_hasher(data.split_at(total_length).0);

    let passed = result == reference;
    if !passed {
        println!("- FAILED: {:?}, {:?}", reference, result)
    }
    passed
}
